sds/src/scanner/mod.rs

```diff
@@ -337,36 +422,71 @@ impl Scanner {
             debug_assert!(content.is_char_boundary(mutated_utf8_match_end));
 
             let matched_content = &content[mutated_utf8_match_start..mutated_utf8_match_end];
+            matched_content_copy = Some(matched_content.to_string());
+        }
 
-            if let Some(replacement) = rule.get_match_action().get_replacement(matched_content) {
-                let before_replacement = &matched_content[replacement.start..replacement.end];
+        if rule.get_match_action().is_mutating() {
+            let mutated_utf8_match_start =
+                (rule_match.utf8_start as isize + *utf8_byte_delta) as usize;
+            let mutated_utf8_match_end = (rule_match.utf8_end as isize + *utf8_byte_delta) as usize;
 
-                // update indices to match the new mutated content
-                <E>::adjust_shift(
-                    custom_index_delta,
-                    before_replacement,
-                    &replacement.replacement,
-                );
-                *utf8_byte_delta +=
-                    replacement.replacement.len() as isize - before_replacement.len() as isize;
+            // Matches for mutating rules must have valid indices
+            debug_assert!(content.is_char_boundary(mutated_utf8_match_start));
+            debug_assert!(content.is_char_boundary(mutated_utf8_match_end));
+
+            let matched_content = &content[mutated_utf8_match_start..mutated_utf8_match_end];
+
+            if rule.get_match_action().is_mutating() {
```

Comment:
I think this condition will always be true

Reply:
why would it be ? get_match_action can be = NoAction

Reply:
If I read the code correctly it is checked line 439 and 428

Reply:
this should be fixed